File: Managed Source/Arsenal.ImageMounter/Reflection/EquatableBox.vb
Namespace Reflection

    Public NotInheritable Class EquatableBox(Of T As {Structure, IEquatable(Of T)})
        Implements IEquatable(Of T), IEquatable(Of EquatableBox(Of T))

        Public Property Value As T

        Public Sub New()
        End Sub

        Public Sub New(value As T)
            _Value = value
        End Sub

        Public ReadOnly Property HasDefaultValue As Boolean
            Get
                Return _Value.Equals(New T)
            End Get
        End Property

        Public Sub ClearValue()
            _Value = New T
        End Sub

        Shared Widening Operator CType(value As T) As EquatableBox(Of T)
            Return New EquatableBox(Of T)(value)
        End Operator

        Shared Widening Operator CType(box As EquatableBox(Of T)) As T
            Return box._Value
        End Operator

        Public Overrides Function ToString() As String
            Return _Value.ToString()
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return _Value.GetHashCode()
        End Function

        Public Overloads Function Equals(other As EquatableBox(Of T)) As Boolean Implements IEquatable(Of EquatableBox(Of T)).Equals
            Return _Value.Equals(other._Value)
        End Function

        Public Overloads Function Equals(other As T) As Boolean Implements IEquatable(Of T).Equals
            Return _Value.Equals(other)
        End Function

        Public Overloads Overrides Function Equals(obj As Object) As Boolean
            If TypeOf obj Is EquatableBox(Of T) Then
                Return _Value.Equals(DirectCast(obj, EquatableBox(Of T))._Value)
            ElseIf TypeOf obj Is T Then
                Return _Value.Equals(DirectCast(obj, T))
            Else
                Return MyBase.Equals(obj)
            End If
        End Function

    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/Reflection/EquatableBox.cs
//  
//  Copyright (c) 2012-2022, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// 

using System;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Arsenal.ImageMounter.Reflection;

public sealed class EquatableBox<T> : IEquatable<T>, IEquatable<EquatableBox<T>> where T : struct, IEquatable<T>
{
    public T Value { get; set; }

    public EquatableBox()
    {
    }

    public EquatableBox(T value)
    {
        Value = value;
    }

    public bool HasDefaultValue => Value.Equals(default);

    public void ClearValue() => Value = new T();

    public static implicit operator EquatableBox<T>(T value) => new(value);

    public static implicit operator T(EquatableBox<T> box) => box.Value;

    public override string? ToString() => Value.ToString();

    public override int GetHashCode() => Value.GetHashCode();

    public bool Equals(EquatableBox<T>? other) => Value.Equals(other?.Value);

    public bool Equals(T other) => Value.Equals(other);

    public override bool Equals(object? obj)
    {
        if (obj is EquatableBox<T> box)
        {
            return Value.Equals(box.Value);
        }
        else
        {
            return obj is T value
                ? Value.Equals(value)
                : base.Equals(obj);
        }
    }
}